theano/tensor/nnet/nnet.py

```diff
@@ -2049,3 +2050,134 @@ def relu(x, alpha=0):
+
+    If target is specified, it will only compute the outputs of the
+    corresponding targets. Otherwise, if target is None, it will compute all
+    the outputs.
+
+    The outputs are grouped in the same order as they are initially defined.
+
+    Arguments:
+    ----------
+    x: tensor of shape (batch_size, number of features)
+        the minibatch input of the two-layer hierarchical softmax.
+
+    batch_size: int
+        the size of the minibatch input x.
+
+    n_outputs: int
+        the number of outputs.
+
+    n_classes: int
+        (optional, default None)
+        the number of classes of the two-layer hierarchical softmax. It
+        corresponds to the number of outputs of the first softmax. It can be
+        set to None, see the note at the end of the docstring.
+
+    n_outputs_per_class: int
+        (optional, default None)
+        the number of outputs per class. It can be set to None, see the note
+        at the end of the docstring.
+
+    W1: tensor of shape (number of features of the input x, number of classes)
+        the weight matrix of the first softmax, which maps the input x to the
+        probabilities of the classes.
+
+    b1: tensor of shape (number of classes,)
+        the bias vector of the first softmax layer.
+
+    W2: tensor of shape (number of classes, number of features of the input x,
+        number of outputs per class)
+        the weight matrix of the second softmax, which maps the input x to
+        the probabilities of the outputs.
+
+    b2: tensor of shape (number of classes, number of outputs per class)
+        the bias vector of the second softmax layer.
+
+    target: tensor of shape either (batch_size,) or (batch_size, 1)
+        (optional, default None)
+        contains the indices of the targets for the minibatch
+        input x. For each input, the function computes the output for its
+        corresponding target. If target is None, then all the outputs are
+        computed for each input.
+
+    Notes
+    -----
+    n_outputs_per_class and n_classes do not need to be defined. If
+    both are not defined, then they are set to the square root of the
+    number of outputs, which is the most computational efficient
+    configuration. If only one is defined, the other is inferred so that
+    their product equals the number of outputs n_outputs (more precisely it is
+    the smallest integer such that their product is greater or equal to
+    n_outputs).
+
+    """
+
+    # In case one or both of n_outputs_per_class and n_classes are not defined
+    if not n_outputs_per_class and not n_classes:
+        n_outputs_per_class = numpy.ceil(numpy.sqrt(n_outputs))
+        n_classes = numpy.ceil(n_outputs / n_outputs_per_class)
+    elif n_outputs_per_class and not n_classes:
+        n_classes = numpy.ceil(n_outputs / n_outputs_per_class)
+    elif n_classes and not n_outputs_per_class:
+        n_outputs_per_class = numpy.ceil(n_outputs / n_classes)
```

Comment:
@bouthilx Actually I realize that these cases don't make sense... The sizes are defined by ``W1`` and ``W2``, so it is the user who should be careful. I will update this tomorrow.